lib/Fencer/Types.hs

```diff
@@ -152,19 +152,29 @@ data DomainDefinition = DomainDefinition
     deriving stock (Eq, Show)
 
 -- | Config for a single rule tree.
-data DescriptorDefinition = DescriptorDefinition
+data DescriptorDefinition
+  =
+    -- | An inner node with no rate limit
+    DescriptorDefinitionInnerNode
     { descriptorDefinitionKey :: !RuleKey
     , descriptorDefinitionValue :: !(Maybe RuleValue)
-    , descriptorDefinitionRateLimit :: !(Maybe RateLimit)
-    , descriptorDefinitionDescriptors :: !(Maybe [DescriptorDefinition])
+    , descriptorDefinitionDescriptors :: ![DescriptorDefinition]
+    }
+  |
+    -- | A leaf node with a rate limit
+    DescriptorDefinitionLeafNode
+    { descriptorDefinitionKey :: !RuleKey
+    , descriptorDefinitionValue :: !(Maybe RuleValue)
+    , descriptorDefinitionRateLimit :: !RateLimit
```

Comment:
I'm personally not a fan of partial fields. In this case I think I'd probably abstract keys and values out and define this data type as follows:

```haskell
data DescriptionContent
    = DescriptionContentDescriptors  ![DescriptorDefinition]
    | DescriptionContentRateLimit !RateLimit

data DescriptionDefinition a = DescriptionDefinition
    { descriptorDefinitionKey :: !RuleKey
    , descriptorDefinitionValue :: !(Maybe RuleValue)
    , descriptorDefinitionDescriptors :: !DescriptionContent
    }
```

I believe this is a better definition.

Reply:
The old version (i.e., the present in `master`) allows specifying a descriptor that has both a rate limit and a list of sub-descriptors. This doesn't make sense in the domain logic. Your proposal suffers from the same problem, and I'd argue it makes it even worse - it makes it mandatory to specify values that don't make sense in the domain logic. Or am I missing something?

Speaking of partial fields in general, is this idea then incompatible with the idea of sum types?

Reply:
I actually wondered about this too.

> This doesn't make sense in the domain logic. Your proposal suffers from the same problem

Hmm, I don't see how? A leaf descriptor would be defined as,
`DescriptionDefinition ruleKey (Just ruleVal) (DescriptionContentRateLimit rateLimit)`

and a non-leaft descriptor would be defined as  
`DescriptionDefinition ruleKey (Just ruleVal) (DescriptionContentDescriptors [...])`

Right?

Reply:
> The old version (i.e., the present in master) allows specifying a descriptor that has both a rate limit and a list of sub-descriptors. This doesn't make sense in the domain logic. Your proposal suffers from the same problem

No, it doesn't. My version is isomorphic to yours. Note that `DescriptionContent` is a sum type just like your `DescriptionDefinition`, I just extracted common data from the constructors and encoded the choice between descriptors and rate limit as a dedicated data type.

Reply:
Oh, my apologies, @effectfully! I wasn't reading carefully enough and I thought that your `DescriptionContent` is a product type when in fact it is a sum type.